sway/ipc-server.c

```diff
@@ -322,6 +331,203 @@ void ipc_get_pixels(wlc_handle output) {
 	ipc_get_pixel_requests = unhandled;
 }
 
+static int ipc_selection_data_cb(int fd, uint32_t mask, void *data) {
+	assert(data);
+	struct get_clipboard_request *req = (struct get_clipboard_request *)data;
+
+	if (mask & WLC_EVENT_ERROR) {
+		sway_log(L_ERROR, "Selection data fd error");
+		goto cleanup;
+	}
+
+	if (mask & WLC_EVENT_READABLE || true) {
+		static const int step_size = 512;
+		char *data = NULL;
+		int ret = 0;
+		int current = 0;
+
+		// read data as long as there is data avilable
+		// grow the buffer step_size in every iteration
+		do {
+			if (data == NULL) {
+				data = malloc(step_size);
+			} else {
+				data = realloc(data, current + step_size);
+			}
+
+			ret = read(fd, data + current, step_size - 1);
+			if (ret < 0) {
+				sway_log_errno(L_ERROR, "Reading from selection data fd failed");
+				goto cleanup;
+			}
+
+			current += ret;
+		} while (ret == step_size - 1);
```

Comment:
as long as there is any data to read from the fd the wayland event loop will probably call this function in a loop so that this would only introduce additional complexity without making it more secure i believe. Shouldn't the 8mb data limit be enough? the data is only sent once so such an attack could only happen from the ipc side (request loop) in which case a malicious client got access to the clipboard ipc feature which should not happen in the first place...